Add docs on normalization naming rules (#2576)
airbytehq/airbyte: docs/architecture/basic-normalization.md

);
```

### Naming Collisions for un-nested objects 

When extracting nested objects or arrays, the Basic Normalization process needs to figure out new names for the expanded tables.

For example, if we had a `cars` table where a nested column `cars` refers to other similar cars. 

```javascript
{
  "make": "alfa romeo",
  "model": "4C coupe",
  "cars": [
    { "make": "audi", "model": "A7" },
    { "make" : "lotus" , "model":  "elise" }
    { "make" : "chevrolet" , "model":  "mustang" }
  ]
}
```

The expanded table would have a conflict in terms of naming since both are named `cars`.
To resolve this sort of name collisions and ensure a more consistent naming scheme, Basic Normalization is choosing the expanded name as follows:
- `cars` for the original parent table
- `cars_da3_cars` for the expanded nested columns following this naming scheme in 3 parts: `<Parent prefix>_<Hash>_<nested column name>`

> 1. Parent prefix: The entire json path string with '_' characters used as delimiters to reach the parent table name

Reviewer: Should json path = json path to the nested column to be clearer?

2. Hash: Hash of the entire json path to reach the nested column reduced to 3 characters. This is to make sure we have a unique name (in case part of the name gets truncated, see below)
3. Nested column name: name of the column being expanded into its own table

By following this strategy, nested columns should "never" collide with other table names. 
If it does, an exception will probably be thrown either by the normalization process or by DBT that runs afterward.

```sql
CREATE TABLE "cars" (
    "_airbyte_cars_hashid" VARCHAR,
    "_airbyte_emitted_at" TIMESTAMP_WITH_TIMEZONE,
    "_airbyte_normalized_at" TIMESTAMP_WITH_TIMEZONE,

    "make" VARCHAR,
    "model" VARCHAR
);

CREATE TABLE "cars_da3_cars" (
    "_airbyte_cars_hashid" VARCHAR,
    "_airbyte_cars_foreign_hashid" VARCHAR,
    "_airbyte_emitted_at" TIMESTAMP_WITH_TIMEZONE,
    "_airbyte_normalized_at" TIMESTAMP_WITH_TIMEZONE,

    "make" VARCHAR,
    "model" VARCHAR
);
```

### Naming limitations & truncation

> Note that destinations have limitations in terms of identifiers naming, especially on the number of characters that is used.

Reviewer: ```suggestion
Note that different destinations have various naming limitations, most commonly on how long names can be.
```

> For instance, on Postgres the documentation states:

Reviewer: ```suggestion
For instance, the Postgres documentation states:
```

>  NAMEDATALEN-1 bytes of an identifier; longer names can be written in commands, but they will be truncated. By default, NAMEDATALEN is 64 so the maximum identifier length is 63 bytes

> Other modern data warehouses have much higher limits in terms of authorized name lengths so this should not be affecting us that often.

Reviewer: ```suggestion
Most modern data warehouses have name lengths limits on the longer side, so this should not affect us that often.
```

> However, in the rare cases where these limits should be reached, Basic Normalization will have to resort to fallback rules as follows:

Reviewer: ```suggestion
However, in the rare cases where these limits are reached, Basic Normalization will fallback to the following rules:
```

> 1. No Truncate if under destination's character limits

Reviewer: Should we leave this out since it describes a case where it's under the limit?

2. Truncate only the `Parent prefix` to fit into destination's character limits
3. Truncate the `Parent prefix` to at least the 10 first characters, then truncate the nested column name starting in the middle to preserve prefix/suffix substrings intact (whenever a truncate in the middle is made, two '__' characters are also inserted to denote where it happened) to fit into destination's character limits 

As an example from the hubspot source, we could have the following tables with nested columns:

| Description | Example 1 | Example 2 |
| :--- | :--- | :--- |
| Original Stream Name | `companies` | `deals` |

> | Json path to the nested column | `companies/property_engagements_last_meeting_booked_campaign` | `deals/properties/engagements_last_meeting_booked_medium` |

Reviewer: Some suggestions:
*might be nice to show the truncated bits
* I don't fully understand the prefix truncation, maybe have an example?

Author: The examples do include the prefix truncation

I added some colors in the bigquery results (non-truncated) to highlight the differences from the postgres ones (truncated)
